### PR TITLE
Schema Deconflict

### PR DESCRIPTION
**Pull request by dwincort**
## Description
As discussed in the comments of PR https://github.com/haskell-works/avro/pull/117 (see https://github.com/haskell-works/avro/pull/117#issuecomment-549036337 in particular), schema deconflicting would be a little more sensible if it functioned directly on the schemas and not on avro values.  That is, it would make more sense for schema deconflicting to have a type like
```haskell
deconflict :: Schema -> Schema -> Either String Schema
```
than
```haskell
deconflict :: Schema -> Schema -> T.Value Type -> Either String (T.Value Type)
```
Indeed, acting directly on the schemas allows us to go one step further: it becomes simple to make the deconflicting process lazy, only forcing strictness if the user wants to call `fromAvro` strictly.  Thus, we can actually make deconflicting have  the type:
```haskell
deconflict :: Schema -> Schema -> Schema
```
This PR introduces such a `deconflict` function.

## Changes
Internally, I needed to slightly extend the type of `Schema` and `Field` to make this possible.  This had knock-on effects on both the strict and lazy versions of `getAvroOf`.  Specifically, for `Schema`, I added  the following constructors:
```haskell
data Type = ...
  | IntLongCoercion
  | IntFloatCoercion
  | IntDoubleCoercion
  | LongFloatCoercion
  | LongDoubleCoercion
  | FloatDoubleCoercion
  | FreeUnion { ty :: Type }
  | Panic { ty :: Type, err :: String }
```
The first 6 are for numeric type coercions and are pretty straightforward.  We need to know the original type so that we parse correctly, and then we need to know the reader type so that we do  the right conversion.

The `FreeUnion` constructor is to allow the situation where the writer does not expect a union type but the reader does.  In this case, we parse assuming no union and then inject the value into a "free" singleton union.  (Note that the reverse situation weirdly used to be allowed in deconflicting, but I didn't support it in this version.   It doesn't make a lot of sense to me, although it could be supported with something like `FakeUnion { ty :: Type, options :: Vector Type }`.)

The `Panic` constructor is what allows the Schema to be lazy.  It contains the type that the reader expects to parse (so lazy parsing can continue) but its presence indicates that deconflicting failed.  For lazy parsing, an `Error` value is produced with  the given error message, and for strict parsing, the error message is immediately thrown.

The `Field` type has one new field: `fldStatus  :: FieldStatus` where
```haskell
data FieldStatus =
    AsIs
  | Ignored
  | Defaulted (Ty.Value Type)
```
By default, all fields  have  status `AsIs`, indicating that they are normal fields.  If a field is not present in the writer but is in the reader (with a default value), then the status is set to `Defaulted` with the default (and if no default is present, the field's type is set to `Panic`).  On the other hand, if the field is in the reader but not the writer, then its status is set to `Ignored`.  We keep it present in the schema so that it can be properly parsed, but the parsing functions then ignore the values they found.

### Deprecation
I'm not sure what you guys want to do with the old deconflicting code, but to prevent breakages, I've currently left the modules there with deprecation pragmas.

## Performance
The benchmark suite is not set up to handle benchmarking deconflicting in this way; it produces many avro objects that are then converted to different avro objects via the old deconflicting process.  The new method skips this step entirely by doing a single pass through the schema and then parsing directly into the final avro object.  I'm not sure the best way to compare these methods.

## Looking Further
Having the avro object as an intermediate type used to be necessary for schema deconflicting, and the ability to deconflict (i.e., be backwards compatible) is a really great feature of avro.  With this new deconflicting style, the intermediate avro representation is no longer strictly necessary.  In other words, one could write a custom version of `getAvroOf` and hook it up to `decodeContainerWith` and parse straight to some target value they want.  Basically, that probably means we can write a version of `fromAvro` that goes straight from bytestring to Haskell types, which would be pretty cool!

### CODE REVIEW

**Comment by AlexeyRaga** [comment]
Wow, thanks, great work! I will find time to look at it soon!

> the intermediate avro representation is no longer strictly necessary
This is definitely a big step into the right direction!

**Comment by AlexeyRaga** [comment]
@dwincort I started looking at this PR and realised that with these changes our `Schema` becomes non-trippable from/to JSON. That's it, we can't encode it to JSON and decode into the same value.

I don't know how actually important this property is, but it is a nice property to have, so I need to look closer and think about it.
@TikhonJelvis, you have done a fair amount of work on `Schema`, what do you think?

Alternatively we could maybe have this extra information, that is only needed for deconflicting into another data type? Such that deconflicting would result into some `Either Error (Schema, Fixes)`? 
It would complicate `FromAvro` a bit, like instead of being
```
fromAvro :: Value Type -> Result a
```
it would be 
```
fromAvro :: Fixes -> Value Type -> Result a
```
but I don't know if it could be a major showstopper...

@newhoggy what do you think?

**Comment by dwincort** [comment]
> @dwincort I started looking at this PR and realised that with these changes our Schema becomes non-trippable from/to JSON. That's it, we can't encode it to JSON and decode into the same value.

Hmm, is this simply because I didn't extend the `parseSchemaJSON` and `schemaToJSON` functions with the new data I added to schemas?  If so, I believe that's an easy (backwards-compatible) fix.

If this is about the fact that there are now schemas for which the before and after of a decode/encode round-trip _on a value_ are no longer identical, then yes, perhaps that's worth thinking about.  If we extend schemas to allow for deconflicting, then there will no longer be a one-to-one mapping between schemas and avro values (that is, the `Value` type).  But, that's the point, right?  That's what deconflicting is all about: the fact that there _are_ multiple ways to turn the bytes into a value.

Your suggestion to change `fromAvro` seems to stem from this issue of the one-to-one mapping: if `Value` no longer maps one-to-one to the bytes, then rather than extending the schema, we can add `Fixes` to `fromAvro` to reconstruct that mapping.  This does make the lack of bijectivity more obvious (an extra argument is needed), but now it means that `fromAvro` and `toAvro` no longer round-trip (unless `Fixes` is empty or something).  Perhaps an alternative is to embrace the idea that values of `Value` no longer uniquely identify their schema.  Maybe there should be a way to provide a schema when encoding that will override default behavior (for instance, telling an `Int` to be encoded as a `Float`).

When I made the new `deconflict` function, I thought about what it would mean to provide a binary operator to the `Schema` type.  Can `deconflict` be called multiple times?  If so, it's certainly not commutative, but is it associative?  Is `Schema` a semigroup now?  Or some sort of partial semigroup?  I basically punted on all of these questions: `deconflict` is an internal, somewhat advanced feature that really only ever needs to be called once (at the time of decoding), so I decided not to worry too much.  I _think_ it's associative, but I would say doing more than one call to `deconflict` or using a deconflicted schema more than just for decoding should probably be "unsupported behavior".

**Comment by AlexeyRaga** [comment]
> is this simply because I didn't extend the parseSchemaJSON and schemaToJSON functions with the new data I added to schemas?

My concern is that if we _do_ extend `FromJSON`/`ToJSON`, then our JSON representation for Schema will have extra properties that are not defined by Avro specification.
For example, I am not sure if we could send such a JSON to services like Schema Registry and likewise.
And if we don't extend it, then the round trip is broken.

> but now it means that fromAvro and toAvro no longer round-trip (unless Fixes is empty or something). 

Yes, I guess that there could be empty `Fixes`, or perhaps it can even have a `Monoid` instance (for example, if we define it similarly to `newtype Fixes = Fixes (Map FieldPath ExtraInfo)`)

**Comment by dwincort** [comment]
> My concern is that if we do extend `FromJSON`/`ToJSON`, then our JSON representation for Schema will have extra properties that are not defined by Avro specification.

Ahhh, I understand now.  Perhaps it would make more sense to treat deconflicted schemas as a special type: essentially use a newtype wrapper (e.g., `newtype DeconflictedSchema = DeconflictedSchema Schema`) to keep them separate and distinct from regular schemas.  After all, deconflicted schemas "should" only be used for parsing data (they can't even be reliably used to go in the other direction due to sum types).

This is similar to the idea of having a paired `Fixes` type (consider that `DeconflictedSchema` could be defined as `(Fixes, Shema)`), but I actually think `Fixes` will be pretty painful to implement.  Keep in mind that it's not just fields that can be different but also numeric types and sum types, and although your suggestion for `FieldPath` over `Field` may contain extra differentiating information about fields, it may be annoying to keep track of.  Also, `Fixes` seems to still require an intermediate avro type, preventing a future where we can go directly from bytes to Haskell.

So, perhaps we _should_ leave the JSON instances as they are, we mark that the new coercions and error types are not properly supported Schema components, and we make clear (through documentation or something like a newtype) that deconflicted schemas are special.  How does that sound?

Basically what I'm proposing comes down to something like:
```haskell
-- Constructor only exposed in some internal module
newtype DeconflictedSchema = DeconflictedSchema { getDeconflictedSchema :: Schema }
  deriving (...) -- Not deriving ToJSON or FromJSON

-- Member function only exposed in some internal module
class IsSchema a where
  getSchema :: a -> Schema

instance IsSchema Schema where
  getSchema = id
instance IsSchema DeconflictedSchema where
  getSchema = getDeconflictedSchema

deconflict :: (IsSchema s1, IsSchema s2) => s1 -> s2 -> DeconflictedSchema
decodeContainerWithSchema :: (IsSchema s, FromAvro a) => s -> ByteString -> [[a]]
```

**Comment by dwincort** [comment]
Just pinging this to see if there's any further thoughts.

I came across a case at work where it would be great if we could see whether two schemas were compatible (i.e., if they could be deconflicted) even though we have no actual values to deconflict.  The approach in this PR (or something similar) would let us get a deconflicted schema, and by scanning it for `Panic`, we could see if the schema were valid or not.

**Comment by AlexeyRaga** [comment]
@dwincort Sorry, it is a busy time of the year for me now.
I experimenting on your branch (slowly) and investigating if we can fit the extended schema into an Avro Specification.

The specification says:

```
{"type": "typeName" ...attributes...}
where typeName is either a primitive or derived type name, as defined below. Attributes not defined in this document are permitted as metadata, but must not affect the format of serialized data.
```

and I thought that we could indeed serialise/deserialise extended schema by adding extra attributes, but now I think that it can indeed be treated as "affecting the format of serialised data" because what we say here is that, for example, `Long` should be read as `Int` from the serialised bytes.
And these options should only affect decoding and should not (and must not) take any effect when data is encoded.

So I had an idea of maybe annotating Schema constructors with some read rules, such as:

```
data ReadLong = LongAsIs | LongFromInt
data ReadDouble = DoubleAsIs | DoubleFromFloat | DoubleFromLong

data Schema
  = Int
  | Long ReadLong
  | Double ReadDouble
```

or maybe use GADTs?

```
data ReadRule a where
  AsIs :: ReadRule a
  LongFromInt :: ReadRule Int64
  DoubleFromFloat :: ReadRule Double
  DoubleFromLong :: ReadRule Double

data Schema
  = Int
  | Long (ReadRule Int64)
  | Double (ReadRule Double)
```

or maybe even GADTs with DataKinds?

```
data ReadRule a where
  AsIs :: ReadRule a
  LongFromInt :: ReadRule 'Schema.Long
  DoubleFromFloat :: ReadRule 'Schema.Double
  DoubleFromLong :: ReadRule 'Schema.Double
```

This way we probably won't need new constructors in `Schema`, we could simply adjust read rules in annotations.

If that works then can we do it as per `Trees that grow` paper, or even with a simplified approach and hide these annotations in "decoration"?

This is what I am trying to do now in your branch: to see if I can avoid having new constructors and serialise this schema to JSON as something like

```
{ "type": { "type": "Long", "readRule": "FromInt" } }
```

and it'd not violate the Avro Specification.
I think that it could also be achieved with extra constructors that you have added, but I'd like to see if it can be nicer with annotations (or not).

One thing that I'd like to change is to get rid of `Panic` constructor.
I would much prefer to make `Schema` strict and to get `Either err Schema` from a deconflicting function. 

Another question that I had was: `FieldStatus` has `Defaulted` constructor. Why do we need it instead of setting the existing `fldDefault` field? If we can do that, then we perhaps don't need `FieldStatus` at all, we could just add `fldReadIgnore` flag to `Field`?

Again, a huge question for me is how to serialise changes in `Field` to JSON. Avro Specification doesn't mention anything about having extra attributes at that level.

**Comment by AlexeyRaga** [comment]
Alternatively we could try making an extended copy of `Schema`, call it `ReadSchema` and only use it internally...
This way we would keep `Schema` Avro Spec compatible, although that would require some code duplication...

**Comment by dwincort** [comment]
> So I had an idea of maybe annotating Schema constructors with some read rules

Clever!  All of those options looks good, although I think I'm partial to the middle one (GADTs without DataKinds).

> One thing that I'd like to change is to get rid of `Panic` constructor.  I would much prefer to make `Schema` strict and to get `Either err Schema` from a deconflicting function.

This is fine by me.  The only reason I added the `Panic` constructor was to support lazy deconflicting, but if you don't want this, I'm happy to cut it.

> Another question that I had was: `FieldStatus` has `Defaulted` constructor. Why do we need it instead of setting the existing `fldDefault` field? If we can do that, then we perhaps don't need `FieldStatus` at all, we could just add `fldReadIgnore` flag to `Field`?

Yup, looking at the code now after being away from it for a bit, `FieldStatus` seems redundant and overkill.  A boolean `fldReadIgnore` flag in `Field` looks like it would be sufficient.

> we could try making an extended copy of Schema, call it ReadSchema and only use it internally

This is very much what I was thinking about with the `DeconflictedSchema` newtype.  It's not exactly my favorite option, but I don't think it would be that much code duplication.

----
I have two more questions to consider.

FreeUnion

writing with the schema

**Comment by AlexeyRaga** [comment]
@dwincort Well, I have tried the annotations approach (https://github.com/haskell-works/avro/pull/127) and I tried two schemas approach. Both went terribly complicated. 

So perhaps your solution is the best middle-ground :)

I think I am willing to compromise on the JSON thing and to keep new constructors in `Schema`.
I will add them to JSON encoding/decoding thing and will ping you on that PR.

I think this excuse can be used: If you serialise a deconflicted Schema to JSON then you'll get a "standard" AVSC JSON with one extra field on `Field` type.
Most likely it won't break anything (since it is just an extra field) and it is highly unlikely that anyone would do it anyway...

**Comment by AlexeyRaga** [comment]
Yes, that works.
I also think now that `schemaToJson` should just serialise these new constructors as their original counterparts:

```
  IntLongCoercion     -> A.String "long"
  IntFloatCoercion    -> A.String "float"
  IntDoubleCoercion   -> A.String "double"
  LongFloatCoercion   -> A.String "float"
  LongDoubleCoercion  -> A.String "double"
  FloatDoubleCoercion -> A.String "double"
  FreeUnion ty        -> schemaToJSON context ty
```

The reason is that I'd like to stay conservative just in case if anyone (future me) gets this "extended" schema and uses it somewhere in a container or in Schema Registry. Then the values that are written with this schema could not be read with this schema anymore.

This is still a problem: now we have a variation of a schema that doesn't obey `fromAvro . toAvro = id` property. This is kind of big, but I don't have a solution to it at the moment.

Now, if you remove `Panic` from `Schema` we can merge this PR and iterate on it.

**Comment by dwincort** [comment]
I removed `Panic` as  a constructor from `Schema`.  Now, the lazy decode/deconflict is not quite as lazy (it will only be lazy if deconflicting works), and the types are very slightly more annoying (`deconflict :: Schema -> Schema -> Either String Schema` instead of just `Schema -> Schema -> Schema`), but getting rid of `Panic` is a pretty nice win.

Would you like to make the `schemaToJSON` changes or should I?

**Comment by dwincort** [comment]
I know I just removed `Panic`, but it occurred to me that if we wanted, we could keep the laziness in this library and have `Panic` be precisely that: an error.  In other words, we could make `schemaToJSON :: Maybe TypeName -> Schema -> Either String A.Value` and define the `ToJSON` instance as:
```haskell
toJSON = either error id . schemaToJSON Nothing
```
This is a little ugly because, presumably, `toJSON` shouldn't throw an error.  But, in some sense, this is no worse than any time one has a lazy value that might throw an error when it is fully evaluated.  Furthermore, we could introduce a function like `strictifySchema :: Schema -> Either String Schema` which is basically like `seq` combined with an error handler.

**Comment by dwincort** [comment]
> Yup, looking at the code now after being away from it for a bit, `FieldStatus` seems redundant and overkill. A boolean `fldReadIgnore` flag in `Field` looks like it would be sufficient.

I just tried implementing this, and I see now why I made `FieldStatus` in the first place.  We need to be able to distinguish the case of simply _having_ a default vs knowing when to use it.  Basically, we have the following cases:
1. The field is in both schemas, and there is no default.  Deconflicting works fine, and we should read the value during decoding.
2. The field is in both schemas, and there is a default.  Deconflicting works fine, and we should read the value during decoding.  The default is ignored (but should still be in the schema).
3. The field is in the reader but not the writer, and there is no default.  Deconflicting fails.
4. The field is in the reader but not the writer, and there is a default.  Deconflicting succeeds, and we should not read the value during decoding.  Use the default instead.
5. The field is in the writer but not the reader (we never use defaults from the writer, so whether there is a default or not doesn't matter).  We should read the value, but we should ignore what it is.

There are 5 cases, but a pair of `(Maybe Type, Bool)` only covers 4 cases.  Cases 1 and 2 are the critical ones that are hard to distinguish with `(Maybe Type, Bool)`.

**Comment by AlexeyRaga** [comment]
>  We need to be able to distinguish the case of simply having a default vs knowing when to use it.

Yes, you are right! Maybe in this case instead of `Bool` it'd be cleaner to use a variation of your initial implementation like

```
data FieldStatus = AsIs | Ignored | Defaulted
``` 

This at least would describe the intent cleaner for future maintenance. We can do it later, too.

Could you please rebase do we can merge?

**Comment by AlexeyRaga** [comment]
I had a thought...

When we eventually remove the `Value` type and do direct encoding/decoding, it could probably look like:

```haskell
class FromEncoding a where
   fromEncoding :: Schema -> ByteString -> Either Text a

instance FromEncoding Int64 where
  fromEncoding Schema.Long bs = getLong bs
  fromEncoding Schema.Int bs = fromIntegral <$> getInt bs
  fromEncoding sch _ = Left ("Unable to decode as Long: " <> sch)
```

With this approach coercion becomes simpler and deconflicting would only need to make sure schema's compatibility (and ignored/defaulted fields).

So perhaps in time we will be able to improve this library quite a lot: better performance, removal of `Value`, removal lazy/strict duplication...
This PR is a first step to a greater future! :)

**Comment by AlexeyRaga** [comment]
I have moved this PR to `schema-deconflict` branch so that we can finish it off there (JSON encoding, etc.).

I will make my own PR(s) to that branch and once it is done will merge to `master`.
